### PR TITLE
[SamuelFoo] iP

### PR DESCRIPTION
**Pull request by SamuelFoo**
# Duck

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duck frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/SamuelFoo/ip/).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is **FREE**!

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the list of commands available represented as an `Enum`:

```java
public enum Command {
    LIST("list"),
    FIND("find"),
    MARK("mark"),
    UNMARK("unmark"),
    DELETE("delete"),
    BYE("bye"),
    TODO("todo"),
    DEADLINE("deadline"),
    EVENT("event");
}
```

### CODE REVIEW

File: src/main/java/Duck.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+                    String response = "Noted. I've removed this task:\n"
+                            + indentText(task.toString(), 2) + "\n"
+                            + String.format("Now you have %s tasks in the list.", TASKS.getTaskCount());
+                    printAsResponse(response);
+                } else if (Command.TODO.equalsName(command)) {
+                    // Tasks
+                    // TODO: Reduce duplicate code
+
+                    String taskPart = lineBuffer.getRemainingLine();
```
Comment (adipanda2002) [inline src/main/java/Duck.java]: Do you think the name taskPart could cause some confusion regarding what 'part' is being referred to? Maybe something like taskDetails might be more relevant in this context.

File: src/main/java/Parser.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        return sj.toString();
+    }
+
+    /**
+     * Read tokens until before the pattern specified has been found
+     * and joins them into a single line. (Excludes the pattern.)
+     *
+     * @return tokens until before the pattern joined into a single line
+     */
+    public String getUntil(String pattern) {
```
Comment (adipanda2002) [inline src/main/java/Parser.java]: What am I getting, and until when? I think the name getUntil could be replaced with something clearer or more descriptive.

File: src/test/java/tasks/DeadlineTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DeadlineTest {
+    @Test
+    public void createDeadline_unmarked_success() throws Exception {
+        // toString
+        Deadline testDeadline1 = new Deadline("test description", new DateAndTime("2024-08-30"));
```
Comment (adipanda2002) [inline src/test/java/tasks/DeadlineTest.java:14]: Could you replace these names that use numbers with more descriptive names that elaborate what type of edge case is being tested?

File: src/main/java/TaskList.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import tasks.Deadline;
+import tasks.Event;
+import tasks.Task;
+import tasks.Todo;
+
+public class TaskList {
+    private final ArrayList<Task> tasks = new ArrayList<>();
+    // TODO: Refactor this
+    // Path relative to project directory
+    private static final String[] dirPathParts = { "store" };
```
Comment (adipanda2002) [inline src/main/java/TaskList.java]: Do you agree that using an abbreviation in dirPathParts might be a bit unclear?

File: src/main/java/Parser.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    /**
+     * Read tokens until before the pattern specified has been found
+     * and joins them into a single line. (Excludes the pattern.) Then,
+     * remove the pattern.
+     *
+     * @return tokens until before the pattern joined into a single line
+     */
+    public String getUntilAndRemovePattern(String pattern) {
```
Comment (adipanda2002) [inline src/main/java/Parser.java]: Would you agree that there's a lot of words in this name, but a lack of clarity as to what the method is doing?

**Review by adipanda2002 — COMMENTED** [review]
LGTM. Generally, your naming is done quite well. There's a few nits with regards to naming here and there that I've commented on that would make the overall clarity of your code even better. Given how well you have used OOP, you could support that with optimizing your naming across the codebase.

File: src/main/java/Duck.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
+import java.util.Scanner;
+
+import exceptions.DeadlineUsageException;
+import exceptions.EventUsageException;
```
Comment (elliot-tang) [inline src/main/java/Duck.java]: great job on organising your exception and task classes into packages! to adhere to the coding standard, perhaps you can consider adding all your other classes to a new package (e.g. the "duck" package) as well?

image
attached the relevant portion of the coding standards for your reference.

File: src/main/java/Duck.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    private static final String TEXT_SEPARATOR = "____________________________________________________________";
+    private static final String TEXT_SEPARATOR_WITH_NEWLINE = TEXT_SEPARATOR + "\n";
+
+    // Standard messages
+    private static final String GREETING = String.format("Hello! I'm %s\n", CHATBOT_NAME)
+            + "What can I do for you?";
+    private static final String GOODBYE = "Bye. Hope to see you again soon!";
+
+    // List to store user inputs
+    private static final TaskList TASKS = new TaskList();
```
Comment (elliot-tang) [inline src/main/java/Duck.java]: great job ensuring that all of your constants have been declared with screaming snake case!

File: src/main/java/TaskList.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    private static final String[] filePathParts = { "store", "tasks.csv" };
+    private final File saveDir = new File(String.join(File.separator, dirPathParts));
+    private static final Storage CSV_HANDLER = new Storage(
+            new File(String.join(File.separator, filePathParts)));
+
```
Comment (elliot-tang) [inline src/main/java/TaskList.java]: well done ensuring that your method name stays attached to the open parenthesis that follows it!

File: src/main/java/Parser.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        this.st = new StringTokenizer(line);
+    }
+
+    /**
+     * Returns true if the buffer has more tokens,
+     * false otherwise.
+     *
+     * @return true if the buffer has more tokens,
+     *         false otherwise
+     */
```
Comment (elliot-tang) [inline src/main/java/Parser.java]: perhaps you can consider capitalising the first word and adding a full stop after your return descriptions, in order to format it like a sentence.

image
attached relevant portion of coding standards, see the last bullet point

File: src/main/java/Duck.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        // Print greeting
+        printAsResponse(GREETING);
+
+        // User input loop
+        while (true) {
+            String line = sc.nextLine();
+            Parser lineBuffer = new Parser(line);
+
+            // "bye" and "list" will only work if they are the only word in the line
+            // e.g. "bye bye" would not work
```
Comment (elliot-tang) [inline src/main/java/Duck.java]: great job on ensuring proper indentation for both code and comments and following egyptian style brackets consistently throughout your code!

**Review by elliot-tang — COMMENTED** [review]
LGTM! just a few nits to fix, but overall great job on adhering to the coding stadards for the course!
